modules/userdirectory-canvas/src/main/java/org/opencastproject/userdirectory/canvas/CanvasUserProviderFactory.java

```diff
@@ -0,0 +1,265 @@
+      throw new ConfigurationException(CACHE_SIZE, "is not set");
+    } else {
+      try {
+        cacheSize = Integer.parseInt(properties.get(CACHE_SIZE).toString());
+
+      } catch (NumberFormatException e) {
+        throw new ConfigurationException(CACHE_SIZE, "is invalid , use a positive integer");
+      }
+    }
+
+    int cacheExpiration;
+    if (properties.get(CACHE_EXPIRATION) == null) {
+      throw new ConfigurationException(CACHE_EXPIRATION, "is not set");
+    } else {
+      try {
+        cacheExpiration = Integer.parseInt(properties.get(CACHE_EXPIRATION).toString());
+
+      } catch (NumberFormatException e) {
+        throw new ConfigurationException(CACHE_EXPIRATION, "is invalid , use a positive integer");
+      }
+    }
+
+    Set<String> instructorRoles;
+    String instructorRoleList = (String) properties.get(CANVAS_INSTRUCTOR_ROLES_KEY);
+
+    if (StringUtils.isNotEmpty(instructorRoleList)) {
+      String[] roles = StringUtils.split(instructorRoleList, " ,");
+      instructorRoles = new HashSet<>(Arrays.asList(roles));
+      logger.debug("Canvas instructor roles: {}", Arrays.toString(roles));
+    } else {
+      throw new ConfigurationException(CANVAS_INSTRUCTOR_ROLES_KEY, "is not set");
+    }
+
+    // Now that we have everything we need, go ahead and activate a new provider, removing an old one if necessary
+    ServiceRegistration existingRegistration = providerRegistrations.remove(pid);
+    if (existingRegistration != null) {
+      existingRegistration.unregister();
+    }
+
+    Organization org;
+    try {
+      org = orgDirectory.getOrganization(organization);
+    } catch (NotFoundException e) {
+      logger.warn("Organization {} not found!", organization);
+      throw new ConfigurationException(ORGANIZATION_KEY, "not found");
+    }
+
+    logger.debug("Creating new CanvasUserProviderInstance for pid={}", pid);
+    CanvasUserProviderInstance provider = new CanvasUserProviderInstance(pid, org, url, token, courseIdentifierProperty,
+        coursePattern, userPattern, userNameProperty, instructorRoles, cacheSize, cacheExpiration);
+
+    providerRegistrations.put(pid, bundleContext.registerService(UserProvider.class.getName(), provider, null));
+    providerRegistrations.put(pid, bundleContext.registerService(RoleProvider.class.getName(), provider, null));
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see org.osgi.service.cm.ManagedServiceFactory#deleted(java.lang.String)
+   */
+  @Override
+  public void deleted(String pid) {
+    logger.debug("Delete CanvasUserProviderInstance for pid={}", pid);
+    ServiceRegistration registration = providerRegistrations.remove(pid);
+    if (registration != null) {
+      registration.unregister();
+      try {
+        ManagementFactory.getPlatformMBeanServer().unregisterMBean(CanvasUserProviderFactory.getObjectName(pid));
+      } catch (Exception e) {
+        logger.warn("Unable to unregister mbean for pid='{}': {}", pid, e.getMessage());
```

Comment:
Add the exception directly to the logger to log this. Don't use `e.printStackTrace();`.
```suggestion
        logger.warn("Unable to unregister mbean for pid='{}'", pid, e);
```